src/main/java/com/atlassian/db/replica/api/AuroraPostgresLsnReplicaConsistency.java

```diff
@@ -0,0 +1,99 @@
+package com.atlassian.db.replica.api;
+
+import com.atlassian.db.replica.internal.MonotonicMemoryCache;
+import com.atlassian.db.replica.spi.Cache;
+import com.atlassian.db.replica.spi.ReplicaConsistency;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * It remembers last write LSN (log sequence number) and compares it with LSN for subsequent reads to detect inconsistencies.
+ * If it cannot remember the LSN of last write, pessimistically assumes it's going to be inconsistent.
+ *
+ * @see <a href="https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-global-database-monitoring.html">Monitoring Aurora PostgreSQL-based Aurora global databases</a>
+ * <p>
+ * It's an PostgreSQL-based Aurora database engine specific implementation.
+ */
+public final class AuroraPostgresLsnReplicaConsistency implements ReplicaConsistency {
+
+    private final Cache<Long> lastWrite;
+
+    public static final class Builder {
+        private Cache<Long> lastWrite = new MonotonicMemoryCache<>();
+
+        /**
+         * @param lastWrite remembers last write
+         */
+        public Builder cacheLastWrite(Cache<Long> lastWrite) {
+            this.lastWrite = lastWrite;
+            return this;
+        }
+
+        /**
+         * @return consistency assuming that LSN (log sequence number) is greater or equal to LSN for last write (if known)
+         */
+        public AuroraPostgresLsnReplicaConsistency build() {
+            return new AuroraPostgresLsnReplicaConsistency(lastWrite);
+        }
+    }
+
+    private AuroraPostgresLsnReplicaConsistency(Cache<Long> lastWrite) {
+        this.lastWrite = lastWrite;
+    }
+
+    @Override
+    public void write(Connection main) {
+        try {
+            lastWrite.put(queryMainDbLsn(main));
+        } catch (Exception e) {
+            throw new RuntimeException("failure during LSN fetching for main database", e);
+        }
+    }
+
+    @Override
+    public boolean isConsistent(Supplier<Connection> replica) {
```

Comment:
Let's try to make it more concise/idiomatic?

Reply:
We could have this, but exceptions are not helping. Not sure if inlining exception handling to `queryReplicaDbLsn` makes it more readable.

```suggestion
    public boolean isConsistent(Supplier<Connection> replica) {
        return lastWrite.get()
            .map(lw -> queryReplicaDbLsn(replica.get()) >= lw)
            .orElseGet(false);
    }
```

Reply:
To unblock functional style, we'd have to get rid of the checked exceptions from `queryReplicaDbLsn`. We could convert checked -> unchecked there.

Reply:
Maybe this 🤔 
```suggestion
        public boolean isConsistent(Supplier<Connection> replica) {
        try {
            long lsn = queryReplicaDbLsn(replica.get());

            return lastWrite.get()
                .map(lw -> lsn >= lw)
                .orElse(false);
        } catch (SQLException e) {
            return false;
        }
    }
```

Reply:
Look that we don't need to fetch replica LSN when there is not LSN for lastWrite, so this version has worse performance...

Reply:
Ok, it looks good:
`try {
            return lastWrite.get()
                .map(lastWriteLsn -> queryReplicaDbLsn(replica.get()) >= lastWriteLsn)
                .orElse(false);
        } catch (Exception e) {
            return false;
        }`
        
WDYT?